MovingBlocks/Terasology
# [WIP] Zones

### Contains

This is the first implementation of zones (described in [this forum post](http://forum.terasology.org/threads/new-conceptual-layer-sector-plus-musings-on-multi-world-node.1420/#post-12708) and lower down in that thread).

### Notes

It implements zones at the world generation level, so they can be used to specify different zones for generation and preview (e.g. a surface zone, and an underground zone). In the world preview, the desired zone can be selected, and that one will be previewed. Each zone can determine its own preview image (via FacetLayers), which just uses the current preview system (so they can be created and customized in the same way).

The region selection is currently done with a function that takes the world position of the block, and the Region (so facets can be accessed), and returns a boolean of whether or not that block is in the region. This is very customizable, but I'm not sure about the performance of it (a more restrictive criteria for determining what is in the zone might allow for much greater optimization, at the chunk and the block level (e.g. being able to ignore whole chunks at a time)).

The current implementation does rasterization by saving and restoring all blocks that aren't in the given zone, which works for now but probably has lots of room for optimization.

I did get into a bit of difficulty with the current way some things are implemented with annotations. I wanted to make the zone itself the `FacetProvider` for all of the facets it contains, but because `FacetProvider`s have to be declared with annotations, I don't think that's possible at the moment (or, at least, I don't think that's how they're intended to be used, and it might be quite fragile). I got around it just fine by adding the zone's list of providers rather than the zone itself, but I'm not sure if that limits some things that the zone can do because it can't easily access the facets before they get added (for example, when zone connectors are added).

### How to test

The [PrototypeWorld](https://github.com/Vizaxo/PrototypeWorld/tree/v2/zones) `v2/zones` branch has the world generator (`ProtoGenerator.java`) set up to use zones. It has an underground zone, a surface zone, and an example zone which uses an odd region function (currently without a preview layer, so it should probably be removed from the preview zones list as it just appears black). 
In practice the example zone would be better off implemented with a facet, but this is just an example of how zones can work.

### Outstanding before merging

This is just a WIP, it's not ready for merging yet.
